### ISSUE TITLE
(xunit.v3) review usage of GetAwaiter().GetResult() in auto-generated entrypoint

### ISSUE THREAD

**Issue by sliekens** [xunit/xunit#3105]
Hello,

I'm upgrading my tests from xUnit.net 2 to 3 and now I have a test that deadlocks in v3 (it ran fine in v2).

I think the problem is the use of `GetAwaiter().GetResult()` in the generated Main entrypoint.

Simple repro: a test which counts down from 10 to 0 and asserts that 10 seconds have passed. Note that I'm using `System.Threading.Timer`, which has a particularly difficult `Dispose` implementation that requires the use of a `WaitHandle`

I poked around in VS2022 threads view and I think these two lines of code are deadlocking each other.

`WaitHandle.WaitOne()` in my code:

```csharp
callbacksFinished.WaitHandle.WaitOne();
```

`GetAwaiter().GetResult()`, which is known for causing deadlocks, in the autogenerated code:

```csharp
return global::Xunit.Runner.InProc.SystemConsole.ConsoleRunner.Run(args).GetAwaiter().GetResult();
```

Full repro:

```csharp
using System.Diagnostics;

namespace xunit3;

public class UnitTest1
{
    [Fact]
    public async Task Test1()
    {
        var stopwatch = Stopwatch.StartNew();
        using var sut = new Countdown(10);

        await sut.Alarm;
        Assert.True(stopwatch.ElapsedMilliseconds >= 10000);
    }
}

public sealed class Countdown : IDisposable
{
    public int TimeLeft { get; private set; }

    public Task Alarm;

    private bool _disposed;

    private readonly Timer _timer;

    public Countdown(int seconds)
    {
        var taskCompletionSource = new TaskCompletionSource();
        _timer = new(
            state => Tick((TaskCompletionSource)state!),
            taskCompletionSource,
            TimeSpan.FromSeconds(1),
            TimeSpan.FromSeconds(1)
        );

        TimeLeft = seconds;
        Alarm = taskCompletionSource.Task;
    }

    private void Tick(TaskCompletionSource taskCompletionSource)
    {
        TimeLeft--;

        if (TimeLeft <= 0)
        {
            taskCompletionSource.TrySetResult();
        }
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }

        // Disposing Timer is a bit tricky because a callback might be running at the moment
        // The solution is to pass a WaitHandle which the Timer will set when all callbacks have finished
        using ManualResetEventSlim callbacksFinished = new(false);


        // When Dispose returns true, wait for callbacks to finish and then do the cleanup
        // When Dispose returns false, it means Dispose was called twice, so do nothing
        if (_timer.Dispose(callbacksFinished.WaitHandle))
        {
            // Block until all callbacks have finished
            callbacksFinished.WaitHandle.WaitOne();

            // (not shown: cleanup of other resources happens now)

            _disposed = true;
        }
        else
        {
            // Dispose was called twice, so do nothing
        }
    }
}
```

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <OutputType>Exe</OutputType>
    <RootNamespace>xunit3</RootNamespace>
    <TargetFramework>net8.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="xunit.runner.json" CopyToOutputDirectory="PreserveNewest" />
  </ItemGroup>

  <ItemGroup>
    <Using Include="Xunit" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.12.0" />
    <PackageReference Include="xunit.v3" Version="1.0.0" />
    <PackageReference Include="xunit.runner.visualstudio" Version="3.0.0" />
  </ItemGroup>

</Project>

```

**Comment by bradwilson** [comment]
In attempting to validate your analysis, I did the following:

- Add `<XunitAutoGeneratedEntryPoint>false</XunitAutoGeneratedEntryPoint>` to the `PropertyGroup`

- Create my own `Main` entry point which uses `await`:

  ```csharp
  using Xunit.Runner.InProc.SystemConsole;

  public class Program
  {
      public static async Task<int> Main(string[] args)
      {
          return await ConsoleRunner.Run(args);
      }
  }
  ```

Your repro test still hangs, which would suggest the deadlock is not caused by the `Main` function but rather the (as you say) tricky nature of `Timer.Dispose`.

**Comment by bradwilson** [comment]
My analysis is that you're deadlocking yourself in your `Dispose` method. Switching from `Dispose` to `DisposeAsync` and using `Task.Yield()` to yield your own thread seems to fix the issue:

```csharp
public async ValueTask DisposeAsync()
{
    if (_disposed)
    {
        return;
    }

    // Disposing Timer is a bit tricky because a callback might be running at the moment
    // The solution is to pass a WaitHandle which the Timer will set when all callbacks have finished
    using ManualResetEventSlim callbacksFinished = new(false);


    // When Dispose returns true, wait for callbacks to finish and then do the cleanup
    // When Dispose returns false, it means Dispose was called twice, so do nothing
    if (_timer.Dispose(callbacksFinished.WaitHandle))
    {
        while (!callbacksFinished.WaitHandle.WaitOne(0))
            await Task.Yield();

        // (not shown: cleanup of other resources happens now)

        _disposed = true;
    }
    else
    {
        // Dispose was called twice, so do nothing
    }
}
```

**Comment by bradwilson** [comment]
Note that this code is in a very tight loop, since neither `WaitOne` nor `Task.Yield` is taking any delay. You could add a small amount of delay (just a few milliseconds) to `WaitOne`, or you could use `Task.Delay()` rather than `Task.Yield`, so that you would be in a super tight CPU loop.

**Comment by sliekens** [comment]
Oh I didn't know you could do that. I will have another look at the debugger tomorrow, in case I missed something.

**Comment by bradwilson** [comment]
You need to update the unit test from `using var...` to `await using var...` but otherwise I didn't touch anything else (and obviously I verified this with the standard `Main`, rather than the custom one I used to verify it wasn't `Main` causing the deadlock).

**Comment by bradwilson** [comment]
I should also point out, from a code validation perspective, you should probably change the code order in the Countdown constructor to:

```csharp
public Countdown(int seconds)
{
    var taskCompletionSource = new TaskCompletionSource();
    TimeLeft = seconds;
    Alarm = taskCompletionSource.Task;

    _timer = new(
        state => Tick((TaskCompletionSource)state!),
        taskCompletionSource,
        TimeSpan.FromSeconds(1),
        TimeSpan.FromSeconds(1)
    );
}
```

Your timer callback can in theory be called as soon as the `Timer` object is new'd up here, but the callback depends on both `TimeLeft` and `Alarm` being properly set. That means you should set them _before_ you create the `Timer` object, or else you have a (very unlikely but still possible) race condition.

**Comment by sliekens** [comment]
I'm still trying to understand why this wasn't an issue with xUnit.net 2. You can take my original post and change only the xUnit.net version to 2.9.2 and it will also run to completion.

**Comment by bradwilson** [comment]
The asynchronous machinery in v2 is quite different from v3.

In my mind, it's not that your code was working on v2 and broken on v3, it's that the code has always been broken but luckily you did not hit the deadlock with v2, whereas you seem to be hitting it 100% of the time with v3.

**Comment by sliekens** [comment]
Ok, I will close this as not an issue with xunit, but still wondering why I never hit the deadlock before 💩.

**Comment by sliekens** [comment]
@bradwilson I did eventually track down the cause: `TaskCompletionSource.TrySetResult` runs continuations of the `TaskCompletionSource.Task` on the same thread.

Effectively, my code behaves as if I wrote this:

```diff
private void Tick(TaskCompletionSource taskCompletionSource)
{
    TimeLeft--;

    if (TimeLeft <= 0)
    {
-        taskCompletionSource.TrySetResult();
+        Dispose();
    }
}
```

Since my `Dispose` implementation waits for all callbacks to finish, this is a deadlock.

I still don't know why I never saw this problem in V2.
